plugins/pulp_docker/plugins/registry.py

```diff
@@ -194,34 +231,125 @@ def create_download_request(self, image_id, file_name, destination_dir):
+
+    def create_image_download_request(self, image_tarsum, destination_dir):
         """
-        if self.endpoint:
-            parts = list(urlparse.urlsplit(self.registry_url))
-            parts[1] = self.endpoint
-            return urlparse.urlunsplit(parts)
-        else:
-            return self.registry_url
+        Return a DownloadRequest instance for the given image tarsum.
+        It is desirable to download the actual layer files with a separate
+        downloader (for progress tracking, etc), so we just create the download
+        requests here and let them get processed elsewhere.
+
+        This adds the Authorization header if a token is known for this
+        repository.
+
+        :param image_id:        unique ID of a docker image
+        :type  image_id:        basestring
+        :type  file_name:       basestring
+        :param destination_dir: full path to the directory where file should
+                                be saved
+        :type  destination_dir: basestring
+
+        :return:    a download request instance
+        :rtype:     nectar.request.DownloadRequest
+        """
+        path = self.LAYER_PATH.format(name=self.name, tarsum=image_tarsum)
+        url = urlparse.urljoin(self.registry_url, path)
+        req = DownloadRequest(url, os.path.join(destination_dir, image_tarsum))
+        return req
+
+    def get_manifest(self, reference):
+        """
+        Get the manifest and its digest for the given reference.
+
+        :param reference: The reference (tag or digest) of the Manifest you wish to retrieve.
+        :type  reference: basestring
+        :return:          A 2-tuple of the digest and the manifest, both basestrings
+        :rtype:           tuple
+        """
+        path = self.MANIFEST_PATH.format(name=self.name, reference=reference)
+        headers, manifest = self._get_path(path)
+        return headers['docker-content-digest'], manifest
+
+    def get_tags(self):
+        """
+        Get a list of the available tags in the repository.
+
+        :return: A list of basestrings of the available tags in the repository.
+        :rtype:  list
+        """
+        path = self.TAGS_PATH.format(name=self.name)
+        headers, tags = self._get_path(path)
+        return json.loads(tags)['tags']
+
+    def _get_path(self, path):
+        """
+        Retrieve a single path within the upstream registry, and return a 2-tuple of the headers and
+        the response body.
+
+        :param path: a full http path to retrieve that will be urljoin'd to the upstream registry
+                     url.
+        :type  path: basestring
+
+        :return:     (headers, response body)
+        :rtype:      tuple
+        """
+        url = urlparse.urljoin(self.registry_url, path)
+        _logger.debug(_('Retrieving {0}'.format(url)))
+        request = DownloadRequest(url, StringIO())
+        report = self.downloader.download_one(request)
```

Comment:
I think the version of nectar that supports ``download_one`` might not be the version we ship with pulp 2.6. It's worth confirming. If I'm right, we might consider upgrading the version of nectar shipped with 2.6 rather than re-working this.